docs/en/development/cli-guideline.md

```diff
@@ -0,0 +1,138 @@
+Then add your table headers and rows. 
+
+```java
+      Table table = templateRenderer.createTable();
+      table.addHeader("repoName", "repoType", "repoUrl");
+      for (RepositoryCommandDto dto : dtos) {
+        table.addRow(dto.getNamespace() + "/" + dto.getName(), dto.getType(), dto.getUrl());
+      }
+      templateRenderer.renderToStdout(TABLE_TEMPLATE, ImmutableMap.of("rows", table, "repos", dtos));
+```
+
+#### Result
+```shell
+    NAME                   TYPE URL 
+    scmadmin/nice_repo     git  http://localhost:8081/scm/repo/scmadmin/nice_repo    
+```
+
+### Key/Value Table
+To create a two column (key-value-style) table you can use the `addKeyValueRow()` method.
+
+```java
+    Table table = createTable();
+    RepositoryCommandDto dto = mapper.map(repository);
+    table.addLabelValueRow("repoNamespace", dto.getNamespace());
+    table.addLabelValueRow("repoName", dto.getName());
+    table.addLabelValueRow("repoType", dto.getType());
+    table.addLabelValueRow("repoContact", dto.getContact());
+    table.addLabelValueRow("repoUrl", dto.getUrl());
+    table.addLabelValueRow("repoDescription", dto.getDescription());
+    renderToStdout(DETAILS_TABLE_TEMPLATE, ImmutableMap.of("rows", table, "repo", dto));
+```
+
+#### Result
+
+```shell
+Namespace: scmadmin                                        
+Name     : testrepo                                        
+Type     : git     
+```
+
+## I18n
+The CLI client commands should support multiple languages. 
+This can be done by using translation keys in the related resource bundles. 
+By default, we support english and german translations.
+
+### Example
+```java
+  static final String DEFAULT_TEMPLATE = String.join("\n",
+    "{{repo.namespace}}/{{repo.name}}",
+    "{{i18n.repoDescription}}: {{repo.description}}",
+    "{{i18n.repoType}}: {{repo.type}}",
+    "{{i18n.repoContact}}: {{repo.contact}}"
+  );
+```
+
+The variables starting with `i18n` are translations from the resource bundles. 
+The fields starting with `repo` are context related model data from the repository we are currently accessing.
+
+## Error handling
+There are different options on how to handle errors. 
+You can use the TemplateRender and print the errors or exception messages to stderr channel.
+
+But you also can you throw the exception directly inside your execution. 
+These exceptions will be handled by the CliExceptionHandler and will be printed to the stderr channel based on a specific template.
+
+## Validation
+The CLI commands support Java bean validation. 
+If you want to use this validation you have to inject the CommandValidator and call `validator.validate()` in the first line of the command execution.
+Then you can simply annotate your fields with validation annotations.
+
+### Example
+```java
+  @Email
+  @CommandLine.Option(names = {"--contact", "-c"})
+  private String contact;
+```
+
+```java
+  @Inject
+public MyCommand(CommandValidator validator) {
+  this.validator = validator;
+  }
```

Comment:
```suggestion
```java
  @Inject
  public MyCommand(CommandValidator validator) {
    this.validator = validator;
  }
```